### ISSUE TITLE
Error: Error ensuring Resource Providers are registered.

### ISSUE THREAD

**Issue by hanyang1986** [hashicorp/terraform-provider-azurerm#24804]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.52

### AzureRM Provider Version

3.90

### Affected Resource(s)/Data Source(s)

azurerm

### Terraform Configuration Files

```hcl
required_providers {
    azurerm = ">= 3.0.0"
  }
```

### Debug Output/Panic Output

```shell
Terraform automatically attempts to register the Resource Providers it supports to
│ ensure it's able to provision resources.
│ 
│ If you don't have permission to register Resource Providers you may wish to use the
│ "skip_provider_registration" flag in the Provider block to disable this functionality.
│ 
│ Please note that if you opt out of Resource Provider Registration and Terraform tries
│ to provision a resource from a Resource Provider which is unregistered, then the errors
│ may appear misleading - for example:
│ 
│ > API version 2019-XX-XX was not found for Microsoft.Foo
│ 
│ Could indicate either that the Resource Provider "Microsoft.Foo" requires registration,
│ but this could also indicate that this Azure Region doesn't support this API version.
│ 
│ More information on the "skip_provider_registration" flag can be found here:
│ https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#skip_provider_registration
│ 
│ Original Error: populating Resource Provider cache: listing Resource Providers: loading results: unexpected end of JSON input
│ 
│   with provider["registry.terraform.io/hashicorp/azurerm"],
│   on provider_override.tf line 3, in provider "azurerm":
│    3: provider "azurerm" {
│
```

### Steps to Reproduce

terraform plan

**Comment by rcskosir** [comment]
@hanyang1986 Thank you for taking the time to open this issue. Are you able to provide us with more information regarding your terraform configuration? Another piece of information that would be useful is the logging when this error occurs. This type of logging in Terraform can be enabled by setting the `TF_LOG` environment variable, e.g. `TF_LOG=TRACE terraform apply.`

**Comment by ssawulski** [comment]
> @hanyang1986 Thank you for taking the time to open this issue. Are you able to provide us with more information regarding your terraform configuration? Another piece of information that would be useful is the logging when this error occurs. This type of logging in Terraform can be enabled by setting the `TF_LOG` environment variable, e.g. `TF_LOG=TRACE terraform apply.`

I have observed a similar problem at the same time as reported issue (hence finding out this reported issue)  We have pinned provider version 3.75.  Before I was able to properly investigate, the issue went away, which leads me to suspect a temporary problem on Azure side

**Comment by akorp** [comment]
Looks like the issue is coming from https://github.com/hashicorp/terraform-provider-azurerm/pull/24645 and affects version `3.90.0` which now explicitly requires setting `skip_provider_registration` to true.

This is unfortunate, since now we have to update all our references to `azurerm`, which are many. Until `3.90.0` we were able to register providers in Azure portal and then use `azurerm` without setting `skip_provider_registration` to true. With `3.90.0` our workflows failing with the error above.

**Comment by joshvito** [comment]
We see this as being introduced in 3.90, and persists through 3.93.0.
Our temporary fix is to set provider version to < 3.90 or turn provider registration off.

e.g. 
```
provider "azurerm" {
  features {}

  skip_provider_registration = true
}
```

OR 

```
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "< 3.90.0" 
    }
  }
}
```

**Comment by akorp** [comment]
Another option for us was to register 3 new Resource Providers introduces in version `3.90.0` (https://github.com/hashicorp/terraform-provider-azurerm/pull/24645/files) in the used Azure Subscription.

**Comment by tombuildsstuff** [comment]
@akorp @joshvito 

When running Terraform with limited permissions you _must_ disable the automatic Resource Provider registration, that's an expected ([and documented](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#skip_provider_registration)) requirement - so this is the expected fix rather than a workaround.

Whilst registering these three RP's will workaround this issue, ultimately you still need to disable the automatic Resource Provider registration when running with limited permissions - else you'll hit up against this issue in the future when additional Resource Providers are added to this list.

**Comment by akorp** [comment]
> @akorp @joshvito
> 
> When running Terraform with limited permissions you _must_ disable the automatic Resource Provider registration, that's an expected ([and documented](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#skip_provider_registration)) requirement - so this is the expected fix rather than a workaround.
> 
> Whilst registering these three RP's will workaround this issue, ultimately you still need to disable the automatic Resource Provider registration when running with limited permissions - else you'll hit up against this issue in the future when additional Resource Providers are added to this list.

@tombuildsstuff thanks for the additional input. We considered the suggested option with `skip_provider_registration = true`, however adding this will not help hitting the same issue (unregistered provider and inability to register it with low privileges SP) if some will be added in the future. Secondly the error message will be more cryptic as per ([ documention](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#skip_provider_registration) 
```
however, please note that the error messages returned from Azure may be confusing as a result (example: API version 2019-01-01 was not found for Microsoft.Foo
```
We preferer more clear error message.
```
Could indicate either that the Resource Provider "Microsoft.Foo" requires registration.
```
 So we could not see the benefits of the suggested default solution with `skip_provider_registration = true`.

**Comment by AtzeDeVries** [comment]
> We are getting this error on a service principle that has owner permissions at the subscription level. All of the work is being performed in this subscription.
> 
> I wouldn't classify it as **limited** permissions.

Same here.

**Comment by joshvito** [comment]
tldr; Updated error log output helped us trace the root cause of our issue.

_`│ Original Error: Cannot register providers: Microsoft.DataFactory, Microsoft.AppConfiguration. Errors were: Cannot register provider Microsoft.DataFactory with Azure Resource Manager: unexpected status 403 (403 Forbidden) with error: AuthorizationFailed: The client '<GUID>' with object id 'GUID' does not have authorization to perform action 'Microsoft.DataFactory/register/action' over scope '/subscriptions/GUID' or the scope is invalid. If access was recently granted, please refresh your credentials..`_


Seems like the more recent updates to the AzureRM provider give better error messaging around this issue.

This helped us narrow down our specific problem.
Here is the barebones of our setup to repro the issue.

```
// main.tf

// default service principle(SP) is Owner of this subscription
provider "azurerm" {
  features {}
}

//  a common module we use for AI
module "app_insights" {
  source = "git::https://path.to.module/azurerm_application_insights//?ref=main"
...
}
```

```
// module.app_insights.ts

// The Service Principle running this subscription is NOT an owner (limited permissions)
provider "azurerm" {
  features {}
  // adding this declaration, allows us to skip registration when accessing this secondary subscription
  skip_provider_registration = true 
  alias           = "logAnalyticsWorkspaceProvider"
  subscription_id = local.log_analytics_workspace_config[var.deployment_environment].subscription_id
}

data "azurerm_log_analytics_workspace" "this" {
  ...
  provider            = azurerm.logAnalyticsWorkspaceProvider
}

resource "azurerm_application_insights" "this" {
  ...
  workspace_id        = data.azurerm_log_analytics_workspace.this.id
}
```

We **wrongly assumed**, since the SP was the owner of the subscription where "work" was being done, that the registration of providers in the secondary subscription didn't need to happen. After all, we are just reading an id from that secondary subscription and the SP has the appropriate permissions to be able to do that.

Have a great weekend.

**Comment by marcindulak** [comment]
Have just experienced the errors once. It's the first time I see on a Terraform Cloud workspace that has more or less daily applies. It didn't happen on the next `terraform apply` executed within a few minutes after the error. No permissions were changed in the meantime. Adding it to the list of nondeterministic behaviours at https://github.com/hashicorp/terraform-provider-azurerm/issues/25585

```
...
API version 2019-XX-XX was not found for Microsoft.Foo
│
│ Could indicate either that the Resource Provider "Microsoft.Foo" requires registration,
│ but this could also indicate that this Azure Region doesn't support this API version.
...
```

Using https://github.com/hashicorp/terraform-provider-azurerm/releases/tag/v3.104.2 and https://github.com/hashicorp/terraform/releases/tag/v1.8.5

Is this a proper way to setup an alias?

```
provider "azurerm" {
  features {}
}

provider "azurerm" {
  features {}
  alias           = "some_subscription"
  subscription_id = "XXX"
  }
```

**Comment by codermrrob** [comment]
Agree that this error message is not clear about limited permissions or what permissions are needed. I am a tenant owner and I can raise the necessary permissions, but it is unclear what permissions are needed.

**Comment by k7kaushal** [comment]
Still facing the same issue, tried all above mentioned steps and solutions.
